### PR TITLE
[GLUTEN-7509][VL] Memory management: Release all native memory managers after all Velox tasks were released

### PR DESCRIPTION
**Pull request by zhztheplayer**
Previous release order:

1. Release Velox task A
2. Release memory manager A
3. Release Velox task B
4. Release memory manager B

After the change:
1. Release Velox task A
2. Release Velox task B
3. Release memory manager A
4. Release memory manager B

This is to prevent possible core dumps if e.g., task A's output data is still held by task B when Spark task is ending.

### CODE REVIEW

**Comment by FelixYBW** [comment]
Is there known issue?

**Comment by zhztheplayer** [comment]
> #6736

@zhouyuan This is not for #6736, just renamed the PR back in case of misleading

**Comment by zhouyuan** [comment]
> > #6736
> 
> @zhouyuan This is not for #6736, just renamed the PR back in case of misleading

@zhztheplayer 
Can you please create one ISSUE to track this?

File: cpp/core/memory/MemoryManager.cc [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  memoryManagerFactories().registerObj(kind, std::move(factory));
+}
+
+MemoryManager* MemoryManager::create(const std::string& kind, std::unique_ptr<AllocationListener> listener) {
+  auto& factory = memoryManagerFactories().get(kind);
+  return factory(std::move(listener));
+}
+
+void MemoryManager::release(MemoryManager* memoryManager) {
+  delete memoryManager;
```
Comment (zhouyuan) [inline cpp/core/memory/MemoryManager.cc:41]: seems a dangerous op

File: cpp/core/utils/Registry.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+class Registry {
+ public:
+  void registerObj(const std::string& kind, T t) {
+    std::lock_guard<std::mutex> l(mutex_);
+    GLUTEN_CHECK(map_.find(kind) == map_.end(), "Already registered for " + kind);
+    map_[kind] = std::move(t);
+  }
+
+  T& get(const std::string& kind) {
+    std::lock_guard<std::mutex> l(mutex_);
```
Comment (PHILO-HE) [inline cpp/core/utils/Registry.h:37]: I note these are code refactors. Does read operation deserve to be optimized with `std::shared_lock` to allow concurrency?

File: cpp/core/memory/MemoryManager.h [hunk 3]
```diff
@@ -19,11 +19,17 @@
 
 #include "arrow/memory_pool.h"
 #include "memory.pb.h"
+#include "memory/AllocationListener.h"
 
 namespace gluten {
 
 class MemoryManager {
  public:
+  using Factory = std::function<MemoryManager*(std::unique_ptr<AllocationListener> listener)>;
```
Comment (PHILO-HE) [inline cpp/core/memory/MemoryManager.h:28]: Nit: remove unnecessary `listener`.

File: cpp/core/utils/Registry.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+#include <unordered_map>
+
+#include "utils/Exception.h"
+
+namespace gluten {
+template <typename T>
+class Registry {
+ public:
+  void registerObj(const std::string& kind, T t) {
+    std::lock_guard<std::mutex> l(mutex_);
```
Comment (PHILO-HE) [inline cpp/core/utils/Registry.h:31]: Nit: better to name it `lock` or some other name.

**Review by PHILO-HE — COMMENTED** [review]
Just some minor comments. Thanks!

File: cpp/core/memory/MemoryManager.cc (hunk 1 above)
Comment (zhztheplayer) [reply cpp/core/memory/MemoryManager.cc:41]: Such operations are needed anyway. Their lifecycles have to be managed directly through JNI. We don't have a mechanism yet to propagate shared pointer references to JVM side.

File: cpp/core/utils/Registry.h (hunk 2 above)
Comment (zhztheplayer) [reply cpp/core/utils/Registry.h:37]: No need to do that so far. We can optimize once it's really required.

File: cpp/core/utils/Registry.h (hunk 4 above)
Comment (zhztheplayer) [reply cpp/core/utils/Registry.h:31]: `std::lock_guard` is already explanatory?

**Comment by zhztheplayer** [comment]
@PHILO-HE @zhouyuan

**Review by PHILO-HE — APPROVED** [review]
Thanks!

**Pull request merged**